File: llm-harness-js/src/clients/multimodal.ts
import { readFile } from "node:fs/promises";
import { PathLike } from "node:fs";
import path from "node:path";

type SupportedCategory = "image" | "video" | "audio" | "file" | "text";

const SUPPORTED_EXTENSIONS: Record<string, [SupportedCategory, string]> = {
  ".jpg": ["image", "image/jpeg"],
  ".jpeg": ["image", "image/jpeg"],
  ".png": ["image", "image/png"],
  ".gif": ["image", "image/gif"],
  ".webp": ["image", "image/webp"],
  ".mp4": ["video", "video/mp4"],
  ".mpeg": ["video", "video/mpeg"],
  ".mov": ["video", "video/quicktime"],
  ".webm": ["video", "video/webm"],
  ".mp3": ["audio", "audio/mpeg"],
  ".wav": ["audio", "audio/wav"],
  ".pdf": ["file", "application/pdf"],
  ".txt": ["text", "text/plain"],
  ".md": ["text", "text/markdown"],
};

function encodeBase64(data: Uint8Array): string {
  return Buffer.from(data).toString("base64");
}

function createTextBlock(text: string): Record<string, unknown> {
  return { type: "text", text };
}

function createImageBlock(dataUrl: string): Record<string, unknown> {
  return { type: "image_url", image_url: { url: dataUrl } };
}

function createFileBlock(
  filename: string,
  dataUrl: string,
): Record<string, unknown> {
  return { type: "file", file: { filename, file_data: dataUrl } };
}

function createAudioBlock(
  encodedData: string,
  format: "wav" | "mp3",
): Record<string, unknown> {
  return { type: "input_audio", input_audio: { data: encodedData, format } };
}

export class MediaMessage {
  public readonly role = "user";
  public readonly content: Array<Record<string, unknown>>;

  constructor({
    paths,
    media,
    description = "",
    labelPages = false,
    mimeType = "image/jpeg",
  }: {
    paths?: string | Uint8Array | Array<string | Uint8Array>;
    media?: string | Uint8Array | Array<string | Uint8Array>;
    description?: string;
    labelPages?: boolean;
    mimeType?: string;
  }) {
    const mediaInput = paths ?? media;
    if (!mediaInput) {
      throw new Error("Either 'paths' or 'media' must be provided");
    }

    const items = Array.isArray(mediaInput) ? mediaInput : [mediaInput];
    this.content = [];

    items.forEach((item, idx) => {
      if (labelPages) {
        this.content.push(createTextBlock(`Page ${idx + 1}:`));
      }

      if (item instanceof Uint8Array) {
        this.content.push(...this.fromBytes(item, mimeType));
      } else {
        this.content.push(...this.fromPath(item));
      }
    });

    if (description) {
      this.content.push(createTextBlock(description));
    }
  }

  private fromBytes(
    data: Uint8Array,
    mimeType: string,
  ): Array<Record<string, unknown>> {
    const dataUrl = `data:${mimeType};base64,${encodeBase64(data)}`;
    return [createImageBlock(dataUrl)];
  }

  private fromPath(
    filePath: string | PathLike,
  ): Array<Record<string, unknown>> {
    const normalizedPath = String(filePath);
    const suffix = path.extname(normalizedPath).toLowerCase();
    const supported = SUPPORTED_EXTENSIONS[suffix];

    if (!supported) {
      throw new Error(
        `Unsupported extension: ${suffix}. Supported: ${Object.keys(SUPPORTED_EXTENSIONS).sort().join(", ")}`,
      );
    }

    const [category, mimeType] = supported;

    if (category === "text") {
      throw new Error(
        "Text file support requires async loader; use MediaMessage.fromPathAsync for text media",
      );
    }

    return [
      {
        __deferred_path__: normalizedPath,
        __category__: category,
        __mime_type__: mimeType,
      },
    ];
  }

  static async fromPathAsync(options: {
    paths?: string | Uint8Array | Array<string | Uint8Array>;
    media?: string | Uint8Array | Array<string | Uint8Array>;
    description?: string;
    labelPages?: boolean;
    mimeType?: string;
  }): Promise<MediaMessage> {
    const message = new MediaMessage(options);
    const resolvedContent: Array<Record<string, unknown>> = [];

    for (const block of message.content) {
      if (!("__deferred_path__" in block)) {
        resolvedContent.push(block);
        continue;
      }

      const filePath = String(block.__deferred_path__);
      const category = String(block.__category__) as SupportedCategory;
      const mimeType = String(block.__mime_type__);
      const bytes = await readFile(filePath);

      if (category === "text") {
        resolvedContent.push(createTextBlock(bytes.toString("utf-8")));
        continue;
      }

      const encoded = encodeBase64(bytes);
      const dataUrl = `data:${mimeType};base64,${encoded}`;

      if (category === "image" || category === "video") {
        resolvedContent.push(createImageBlock(dataUrl));
      } else if (category === "file") {
        resolvedContent.push(createFileBlock(path.basename(filePath), dataUrl));
      } else if (category === "audio") {
        resolvedContent.push(
          createAudioBlock(
            encoded,
            path.extname(filePath).toLowerCase() === ".wav" ? "wav" : "mp3",
          ),
        );
      }
    }

    message.content.length = 0;
    message.content.push(...resolvedContent);
    return message;
  }
}

File: llm_harness/clients/multimodal.py
"""Multimodal input utilities for OpenAI Chat Completions API.

Creates content blocks compatible with OpenAI's Chat Completions API format (used by OpenRouter).
"""

import base64
from pathlib import Path
from typing import Any

from langchain_core.messages import HumanMessage

# Supported file types: extension -> (category, mime_type)
SUPPORTED_EXTENSIONS: dict[str, tuple[str, str]] = {
    # Images
    ".jpg": ("image", "image/jpeg"),
    ".jpeg": ("image", "image/jpeg"),
    ".png": ("image", "image/png"),
    ".gif": ("image", "image/gif"),
    ".webp": ("image", "image/webp"),
    # Videos (mapped to image_url for vision models)
    ".mp4": ("video", "video/mp4"),
    ".mpeg": ("video", "video/mpeg"),
    ".mov": ("video", "video/quicktime"),
    ".webm": ("video", "video/webm"),
    # Audio
    ".mp3": ("audio", "audio/mpeg"),
    ".wav": ("audio", "audio/wav"),
    # Documents
    ".pdf": ("file", "application/pdf"),
    # Text
    ".txt": ("text", "text/plain"),
    ".md": ("text", "text/markdown"),
}


def _encode_base64(data: bytes) -> str:
    """Encode bytes to base64 string."""
    return base64.b64encode(data).decode("utf-8")


def _create_text_block(text: str) -> dict[str, Any]:
    """Create text content block for Chat Completions API."""
    return {"type": "text", "text": text}


def _create_image_block(data_url: str) -> dict[str, Any]:
    """Create image content block for Chat Completions API."""
    return {"type": "image_url", "image_url": {"url": data_url}}


def _create_file_block(filename: str, data_url: str) -> dict[str, Any]:
    """Create file content block for Chat Completions API."""
    return {"type": "file", "file": {"filename": filename, "file_data": data_url}}


def _create_audio_block(encoded_data: str, format: str) -> dict[str, Any]:
    """Create audio content block for Chat Completions API."""
    return {"type": "input_audio", "input_audio": {"data": encoded_data, "format": format}}


class MediaMessage(HumanMessage):
    """HumanMessage with media content for Chat Completions API.

    Supports file paths or raw bytes for images. Auto-detects file types from paths.

    Args:
        paths: File path(s) or bytes. Can be a single path/bytes or a list.
        description: Optional text description to append after media.
        label_pages: If True, adds "Page N:" labels before each media item.
        mime_type: MIME type for raw bytes (default: "image/jpeg").

    Example:
        >>> MediaMessage("image.png", "What's in this image?")
        >>> MediaMessage([b"page1", b"page2"], "Describe these pages", label_pages=True)
        >>> MediaMessage(paths=["doc.pdf"], description="Extract text from this PDF")
    """

    def __init__(
        self,
        paths: str | Path | bytes | list[str | Path | bytes] | None = None,
        media: str | Path | bytes | list[str | Path | bytes] | None = None,
        description: str = "",
        label_pages: bool = False,
        mime_type: str = "image/jpeg",
    ):
        # Support both 'paths' and 'media' for backward compatibility
        media_input = paths if paths is not None else media
        if media_input is None:
            raise ValueError("Either 'paths' or 'media' must be provided")

        # Normalize to list
        items = [media_input] if isinstance(media_input, (str, Path, bytes)) else list(media_input)

        content_blocks: list[dict[str, Any]] = []
        for idx, item in enumerate(items, 1):
            blocks = self._from_bytes(item, mime_type) if isinstance(item, bytes) else self._from_path(Path(item))

            if label_pages and blocks:
                content_blocks.append(_create_text_block(f"Page {idx}:"))

            content_blocks.extend(blocks)

        if description:
            content_blocks.append(_create_text_block(description))

        super().__init__(content=content_blocks)

    def _from_bytes(self, data: bytes, mime_type: str) -> list[dict[str, Any]]:
        """Create content blocks from raw bytes (assumes image)."""
        data_url = f"data:{mime_type};base64,{_encode_base64(data)}"
        return [_create_image_block(data_url)]

    def _from_path(self, path: Path) -> list[dict[str, Any]]:
        """Create content blocks from a file path."""
        if not path.exists():
            raise FileNotFoundError(f"File not found: {path}")

        suffix = path.suffix.lower()
        if suffix not in SUPPORTED_EXTENSIONS:
            supported = ", ".join(sorted(SUPPORTED_EXTENSIONS.keys()))
            raise ValueError(f"Unsupported extension: {suffix}. Supported: {supported}")

        category, mime_type = SUPPORTED_EXTENSIONS[suffix]

        if category == "text":
            return [_create_text_block(path.read_text(encoding="utf-8"))]

        # Binary content - read once
        encoded = _encode_base64(path.read_bytes())
        data_url = f"data:{mime_type};base64,{encoded}"

        if category in ("image", "video"):
            return [_create_image_block(data_url)]

        if category == "file":
            return [_create_file_block(path.name, data_url)]

        if category == "audio":
            fmt = "wav" if suffix == ".wav" else "mp3"
            return [_create_audio_block(encoded, fmt)]

        return []
